scripts/train_agents.py

```diff
@@ -2,16 +2,200 @@
+            agent_file_tag = pwadv_tag + '/' + CheckedPoints.FINAL_TRAINED_MODEL)
+        
+        how_long = how_long_init + how_long_for_agent*(round+1)
+        set_input(args=args, quick_test=quick_test, how_long=how_long)
+        args.learner_type = main_agent_type
+        args.reward_magnifier = reward_magnifier
+        args.exp_dir = root_adv + '/' + main_agent_type + '-' + adversary_type + 'play' + '/' + str(round)
+        PwADVs( args=args, 
+                agent_folder_path = root_adv + '/' + main_agent_type + '-' + adversary_type + 'play' + '/' + str(round-1), 
+                agent_file_tag = pwadv_tag + '/' + CheckedPoints.FINAL_TRAINED_MODEL,
+                adv_folder_paths = [root_adv + '/' + adversary_type + '/' + str(round)], 
+                adv_file_tag = adv_tag + '/' + checked_adversary)
+        
+def MultiAdversaryPlay(args, 
+                        exp_tag = 'M2FP', 
+                        main_agent_path = None,
+                        main_agent_type = LearnerType.SUPPORTER, 
+                        adversary_type = LearnerType.SELFISHER, 
+                        checked_adversary = CheckedPoints.FINAL_TRAINED_MODEL, 
+                        how_long_init = 4.0,
+                        how_long_for_agent = 4.0,
+                        how_long_for_adv = 4.0,
+                        rounds_of_advplay = 401,
+                        reward_magnifier = 3.0):
+    
+    args.dynamic_reward = False
+    args.final_sparse_r_ratio = 0.5
+    if main_agent_path is None:
+        how_long = how_long_init
+        set_input(args=args, quick_test=quick_test, how_long=how_long)
+        args.SP_seed = args.ADV_seed
+        args.SP_h_dim = args.ADV_h_dim
+        args.learner_type = main_agent_type
+        args.reward_magnifier = reward_magnifier
+        args.exp_dir = exp_tag 
+        SP(args=args, pop_force_training=True)
+        sp_tag = 'sp_s' + args.SP_seed + '_h' + args.SP_h_dim + '_tr(SP)_ran'
+        main_agent_path = exp_tag + '/' + sp_tag
+    root = 'agent_models/' + main_agent_path
+    root_adv = root + '/' + exp_tag
+
+    adv_tag = 'adv_s' + args.ADV_seed + '_h' + args.ADV_h_dim + '_tr(H)_ran'
+    pwadv_tag = 'pwadv_s' + args.ADV_seed + '_h' + args.ADV_h_dim + '_tr(SP_SPADV)_ran'
+
+    how_long = how_long_for_adv
+    set_input(args=args, quick_test=quick_test, how_long=how_long)
+    args.learner_type = adversary_type
+    args.reward_magnifier = reward_magnifier
+    args.exp_dir = root_adv + '/' + adversary_type + '/0'
+    ADV(args=args, 
+        agent_folder_path = root, 
+        agent_file_tag= CheckedPoints.BEST_EVAL_REWARD)
+
+    how_long = how_long_init + how_long_for_agent
+    set_input(args=args, quick_test=quick_test, how_long=how_long)
+    args.learner_type = main_agent_type
+    args.reward_magnifier = reward_magnifier
+    args.exp_dir = root_adv + '/' + main_agent_type + '-' + adversary_type + 'play' + '/0'
```

Comment:
I also think that `args.exp_dir` is being misused in your code. I suggest appending all these paths to the exp_dir instead of replacing it. Something like: `args.exp_dir = args.exp_dir + '/' + main_agent_type + '-' + ...`

The purpose of args.exp_dir is to have an experiment directory which inside of that are multiple agents of different training types. For example, if I want to test a specific total_training_time, I set args.exp_dir='half-training-times', and then I can run different training methods such as SP, FCP, N-1-SP and etc and all their files will be saved in the directory of 'agent_models/half-training-times'.

Reply:
Good point! But I Need to check multiple functions to see what I can do there. Thank you